tests/test_irq.c

```diff
@@ -0,0 +1,45 @@
+
+void test_irq_reentrancy(TestContext *ctx) {
+	// Make sure that enable_interrupts() does not cause an interrupt
+	// to trigger if called under interrupt.
+	volatile bool cb1_called = false;
+	volatile bool cb2_called = false;
+	bool cb1_running = false;
+	bool fail_reentrant = false;
+	bool fail_order = false;
+
+	void cb1(int ovlf) {
+		cb1_called = true;
+		cb1_running = true;
+		disable_interrupts();
+
+		// Wait for the other timer interrupt to trigger
+		wait_ms(30);
+
+		// This enable_interrupts should not trigger the other
+		// interrupt, otherwise the test will fail.
+		enable_interrupts();
```

Comment:
I am not able to verify why this does not immediately trigger the awating exception. I have tried checking the cause register but the interrupt pending bit for timer is not set just before we re-enable interrupts. What is preventing the secondary interrupt here?

Reply:
it's just that we are already under interrupt (status register, bit 1, EXL).

Reply:
Ah I see, non-maskable interrupts are not subjec to that 👍